Script to validate file names

msafih opened the pull request:
I wrote this script which will validate file and directory names according to our naming convention.

After we merge this, we will be able to run...

```
node ./node_modules/@hollowverse/common-config/validateFileNames.js
```

...from `hollowverse/hollowverse` and other repos...

The output looks like this

![](https://user-images.githubusercontent.com/26730388/29345263-6b3137c4-81f2-11e7-8196-f661df7c1f3c.png)

---

I tried to write this script in TypeScript and I was able to run it fine as...

```
./node_modules/.bin/ts-node validateFileNames.ts
```

...but when I moved it to `@hollowverse/common-config` and tried to run it...

```
./node_modules/.bin/ts-node ./node_modules/@hollowverse/common-config/validateFileNames.ts
```

...I kept getting all kinds of errors. I tried using the `-P` flag to set the right `tsconfig.json` and I tried to remove `node_modules` from the excluded paths in `tsconfig.json`, but no luck.

I'm okay with keeping this in JavaScript and relaxing our guidelines in the wiki a bit more if you're okay with this.

---

On an unrelated note: I changed the GitHub protected branch setting on this repo to not require a review before merging but still require a PR. So, we can send a PR and merge it immediately. That way we create visibility without causing workflow delays. And the option to request a review is still available, so we're good there as well.

I feel some repos, such as this one, might be better off using this workflow as opposed to requiring a review before merging.

forabi commented on validateFileNames.js:
```diff
@@ -0,0 +1,94 @@
+// Now let's go through the files to see if they violate the naming convention
+files.forEach(file => {
+  // Split the path of the file to get the path components
+  const pathComponents = file.split('/');
```
Would that work on Windows? Paths in Windows are split with a backslash (`\`). There must be a Node.js API or a 3rd-party module to get the path components.

Edit: `path.sep` returns the platform-specific separator. [Node.js documentation](https://nodejs.org/api/path.html#path_path_sep) has an example on how to split using `path.sep`.

msafih replied:
Good call. I'll switch to that.

forabi commented on validateFileNames.js:
```diff
@@ -0,0 +1,94 @@
+    );
+  });
+}
+function isCamelCase(filePathComponent) {
```
Should we use [lodash's `camelCase`](https://devdocs.io/lodash~4/index#camelCase) to check if `filePathComponent === camelCase(filePathComponent)` instead of this and the regular expression?

msafih replied:
One issue with lodash's camelCase is that it will strip out the dot, so `camelCase('.gitignore') === '.gitignore'` will be `false`. What do you think?

forabi replied:
I didn't know that. I think it is fine, we probably won't need to check these files (they are usually configuration files for development tools and are all lowercase by convention). I don't know. If we need to we can strip the initial dot before comparing, or we can keep the function as it is now.

msafih replied:
Yeah, I think using `lodash.camelCase` is better. It's smart in the way it handles something like `someFUNKYName` while my regex would actually let that name pass.

msafih replied:
Doesn't seem like `lodash.camelcase` is a good option after all. It converts `webpack.config.js` to `webpackConfigJs`. I'll keep the regex.

forabi commented on validateFileNames.js:
```diff
@@ -0,0 +1,94 @@
+const camelCasedFileNameRegex = /^[.]?([a-z])+([0-9]|[a-zA-Z]|[.])*$/;
+
+// Files and directories that are exempt from the naming convention
+const ignoredFilesAndDirectories = [
```
I think we should use `minimatch` to match patterns of files. It works like this:

```
> minimatch('path/to/Dockerfile', 'Dockerfile', { matchBase: true })
true

> minimatch('customTypings/lodash-modules.d.ts', 'customTypings/*', { matchBase: true })
true

```

Using `minimatch` we can simplify `isIgnored`.

msafih replied:
Okay, I'll look into using `minimatch` to make this more robust. 

Where should I put the dependencies of this script, in the same `package.json` as `common-config`?

forabi replied:
Yes, but as regular (i.e. not peer) dependencies.

msafih replied:
Done

forabi commented:
> ...I kept getting all kinds of errors. I tried using the -P flag to set the right tsconfig.json and I tried to remove node_modules from the excluded paths in tsconfig.json, but no luck.

Yeah, `ts-node` seems to have a lot of problems. I think if we are going to use TypeScript, we should always compile the code before running instead of using `ts-node`. By the way, the value of `-P` should be the path to **the directory containing the `tsconfig.json` file**, not the file itself.

forabi commented:
> I'm okay with keeping this in JavaScript and relaxing our guidelines in the wiki a bit more if you're okay with this.

JavaScript with TypeScript type checker and automatic type acquisition enabled is a good approach for scripts — we get most of the benefits of TS without the dev dependencies and the compile step.

msafih commented on validateFileNames.js:
```diff
@@ -0,0 +1,94 @@
+  'LICENSE.md',
+  'customTypings',
+  'typings',
+];
```
I'm thinking of reading this array of exempt files from a config file that exists in `hollowverse/hollowverse` itself or where ever this script is called from. That will make adding more files convenient, but it will add complexity to the setup of this script. What do you think?

forabi replied:
I had the same idea. We can probably start by checking for the ignore file in `process.cwd()`, if it does not exist, we fall back to this array.

msafih replied:
No, it's fine. This is a fun, easy one, and it's not urgent. I can handle it. 🙂

So, I'm thinking of using `process.cwd()` to look for the config file instead of `find-package-json` for simplicity's sake. Thoughts?

forabi replied:
`process.cwd()` returns the currently process directory, which may not necessarily be the project's root. Suppose I `cd` into `server/src`, and run the command manually there. `process.cwd()` would be `server/src`, not the root directory.

msafih replied:
@forabi I looked into using `find-packag-json`, it's a good idea but it still doesn't provide full protection without additional engineering. Like what if you were to run the script inside `functions/`? There's a `package.json` there, so it would fail.

I think it's fine to go with `process.cwd()` for now.

msafih commented:
@forabi I've addressed the comments.

With regard to using `@ts-check`, I just tried that. It's complaining about not knowing `require`, `shelljs`, `minimatch`, etc. Should I install the `@types` files of those modules? I know you mentioned automatic type acquisition, but that doesn't seem to be a compiler option. (I don't use VS Code.)
